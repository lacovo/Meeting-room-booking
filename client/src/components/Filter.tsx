import React, { useState } from 'react';
import {
  Button,
  TextField,
  Dialog,
  DialogActions,
  DialogContent,
  FormControl,
  InputLabel,
  DialogTitle,
  Select,
  MenuItem,
  InputAdornment,
} from '@material-ui/core';
import { AccountCircle } from '@material-ui/icons';
import { useDispatch, useSelector } from 'react-redux';
import { IAppState, IMatch, IRoom } from '../types';
import { setMatch } from '../redux';
import DateFnsUtils from '@date-io/date-fns';
import {
  KeyboardDatePicker,
  MuiPickersUtilsProvider,
} from '@material-ui/pickers';
import moment from 'moment';

export default function Filter(props: any) {
  const { open, handleClose } = props;
  const dispatch = useDispatch();
  const rooms: IRoom[] = useSelector((state: IAppState) => {
    return state.rooms;
  });
  const match: IMatch = useSelector((state: IAppState) => {
    return state.match;
  });
  const thisDate =
    match.date === '' ? null : moment(match.date, 'YYYY-MM-DD').toDate();
  const [filter, setFilter] = useState({
    roomName: match.roomName,
    date: thisDate,
    host: match.host,
    guest: match.guest,
  });
  const handleSubmitFilter = () => {
    const matchDate =
      filter.date === null ? '' : moment(filter.date).format('YYYY-MM-DD');
    const thismatch = {
      ...match,
      roomName: filter.roomName,
      date: matchDate,
      host: filter.host,
      guest: filter.guest,
    };
    dispatch(setMatch(thismatch));
    handleClose();
  };
  const handleSelect = (eventprops: any) => {
    setFilter({ ...filter, roomName: eventprops.target.value });
  };
  const handleReset = () => {
    const thismatch = { ...match, roomName: '', date: '', host: '', guest: '' };
    setFilter({ roomName: '', date: null, host: '', guest: '' });
    dispatch(setMatch(thismatch));
  };
  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle>Filter</DialogTitle>
      <DialogContent>
        <div>
          <FormControl fullWidth>
            <InputLabel id="demo-simple-select-label">Room Name</InputLabel>
            <Select
              labelId="demo-simple-select-label"
              id="demo-simple-select"
              label="Room"
              value={filter.roomName}
              onChange={handleSelect}
            >
              <MenuItem key={-1} value={''}>All</MenuItem>
              {rooms.map((row: any, index: any) => {
                return <MenuItem key={index} value={row.roomname}>{row.roomname}</MenuItem>;
              })}
            </Select>
          </FormControl>
        </div>
        <div className="filter-div">
          <MuiPickersUtilsProvider utils={DateFnsUtils}>
            <KeyboardDatePicker
              autoOk
              variant="inline"
              inputVariant="outlined"
              label="Select Date"
              format="MM/dd/yyyy"
              value={filter.date}
              InputAdornmentProps={{ position: 'start' }}
              onChange={(date) => setFilter({ ...filter, date: date })}
            />
          </MuiPickersUtilsProvider>
        </div>
        <div className="filter-div">
          <TextField
            id="host"
            label="Host"
            onBlur={(e) => setFilter({ ...filter, host: e.target.value })}
            defaultValue={filter.host}
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  <AccountCircle />
                </InputAdornment>
              ),
            }}
            variant="standard"
          />
        </div>
        <div className="filter-div">
          <TextField
            id="guest"
            label="Guest"
            onBlur={(e) => setFilter({ ...filter, guest: e.target.value })}
            defaultValue={filter.guest}
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  <AccountCircle />
                </InputAdornment>
              ),
            }}
            variant="standard"
          />
        </div>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleReset}>Reset</Button>
        <Button onClick={handleSubmitFilter}>Apply</Button>
      </DialogActions>
    </Dialog>
  );
}
