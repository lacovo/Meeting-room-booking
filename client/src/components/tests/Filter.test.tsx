import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Filter from '../Filter';
import toJson from "enzyme-to-json";

configure({ adapter: new Adapter() });

jest.mock('react-redux', () => ({
  useDispatch: () => {},
  useSelector: () => {
    return [{   status: 0,
      match: {roomId: "", roomName: "", date:"", host: "", guest: "", orderBy: "default", is_all: false},
      bookings: [],
      rooms: [], }];
  },
}));

describe('<Filter />', () => {
  it('Should render and match the snapshot', () => {
    const wrapper = shallow(<Filter />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
