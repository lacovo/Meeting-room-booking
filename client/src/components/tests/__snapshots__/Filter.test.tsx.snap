// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> Should render and match the snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))>
  <WithStyles(ForwardRef(DialogTitle))>
    Filter
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <div>
      <WithStyles(ForwardRef(FormControl))
        fullWidth={true}
      >
        <WithStyles(ForwardRef(InputLabel))
          id="demo-simple-select-label"
        >
          Room Name
        </WithStyles(ForwardRef(InputLabel))>
        <WithStyles(ForwardRef(Select))
          id="demo-simple-select"
          label="Room"
          labelId="demo-simple-select-label"
          onChange={[Function]}
        >
          <WithStyles(ForwardRef(MenuItem))
            key="-1"
            value=""
          >
            All
          </WithStyles(ForwardRef(MenuItem))>
          <WithStyles(ForwardRef(MenuItem))
            key="0"
          />
        </WithStyles(ForwardRef(Select))>
      </WithStyles(ForwardRef(FormControl))>
    </div>
    <div
      className="filter-div"
    >
      <MuiPickersUtilsProvider
        utils={[Function]}
      >
        <PickerWithState
          InputAdornmentProps={
            Object {
              "position": "start",
            }
          }
          allowKeyboardControl={true}
          autoOk={true}
          format="MM/dd/yyyy"
          inputVariant="outlined"
          invalidDateMessage="Invalid Date Format"
          label="Select Date"
          maxDate={2100-01-01T00:00:00.000Z}
          maxDateMessage="Date should not be after maximal date"
          minDate={1900-01-01T00:00:00.000Z}
          minDateMessage="Date should not be before minimal date"
          onChange={[Function]}
          openTo="date"
          value={Date { NaN }}
          variant="inline"
          views={
            Array [
              "year",
              "date",
            ]
          }
        />
      </MuiPickersUtilsProvider>
    </div>
    <div
      className="filter-div"
    >
      <WithStyles(ForwardRef(TextField))
        InputProps={
          Object {
            "startAdornment": <ForwardRef(WithStyles)
              position="start"
            >
              <UNDEFINED />
            </ForwardRef(WithStyles)>,
          }
        }
        id="host"
        label="Host"
        onBlur={[Function]}
        variant="standard"
      />
    </div>
    <div
      className="filter-div"
    >
      <WithStyles(ForwardRef(TextField))
        InputProps={
          Object {
            "startAdornment": <ForwardRef(WithStyles)
              position="start"
            >
              <UNDEFINED />
            </ForwardRef(WithStyles)>,
          }
        }
        id="guest"
        label="Guest"
        onBlur={[Function]}
        variant="standard"
      />
    </div>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      onClick={[Function]}
    >
      Reset
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      onClick={[Function]}
    >
      Apply
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
